[Serialization] Add flag to force unsafe recovery from some xref failures

xymus opened the pull request:
Intro a deserialization mode controlled by the flag `-experimental-force-workaround-broken-modules` to attempt unsafe recovery from deserialization failures caused by project issues.

The one issue handled at this time is when a type moves from one module to another. With this new mode the compiler may be able to pick a matching type in a different module. This is risky to use, but may help in a pinch for a client to fix and issue in a library over which they have no control.

xymus committed:
[Serialization] Add flag to force unsafe recovery from some xref failures

Intro a deserialization mode controlled by the flag
`-experimental-force-workaround-broken-modules` to attempt unsafe
recovery from deserialization failures caused by project issues.

The one issue handled at this time is when a type moves from one module
to another. With this new mode the compiler may be able to pick a
matching type in a different module. This is risky to use, but may help
in a pinch for a client to fix and issue in a library over which they
have no control.

xymus commented:
@swift-ci Please smoke test

bnbarham commented on lib/Serialization/Deserialization.cpp:
```diff
@@ -1945,13 +1946,32 @@ ModuleFile::resolveCrossReference(ModuleID MID, uint32_t pathLen) {
+        errorKind == ModularizationError::Kind::DeclMoved &&
+        !values.empty()) {
+      // Print the error as a remark and notify of the recovery attempt.
+      llvm::handleAllErrors(std::move(error),
+        [&](const ModularizationError &modularError) {
+          modularError.diagnose(this, DiagnosticBehavior::Remark);
+        });
+      getContext().Diags.diagnose(getSourceLoc(),
+                                  diag::modularization_issue_worked_around,
+                                  foundIn);
```
Would it be possible to make this a warning with the remark instead attached as a note?

xymus replied:
Good idea. I inverted both diagnostics, using the warning as the main message with the current remark as a note.

xymus committed:
[Serialization] Make a warning the diagnostic about attempting recovery

Use the `attempting forced recovery` diagnostic as main warning to which
we attach other messages as notes. Also mention the flag in the
diagnostic to reinforce that the flag is active.

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please smoke test
